src/compiler/checker.ts

```diff
@@ -23132,9 +23132,17 @@ namespace ts {
 
         function isConstantReference(node: Node): boolean {
             switch (node.kind) {
-                case SyntaxKind.Identifier:
+                case SyntaxKind.Identifier: {
                     const symbol = getResolvedSymbol(node as Identifier);
-                    return isConstVariable(symbol) || !!symbol.valueDeclaration && getRootDeclaration(symbol.valueDeclaration).kind === SyntaxKind.Parameter && !isParameterAssigned(symbol);
+                    if (isConstVariable(symbol)) {
+                        return true;
+                    }
+                    if (symbol.valueDeclaration) {
+                        const decl = getRootDeclaration(symbol.valueDeclaration);
+                        return isParameter(decl) && !isParameterAssigned(symbol) || isVariableDeclaration(decl) && isCatchClause(decl.parent);

```

Comment:
Catch variables can be reassigned too. What happens with a test like

```ts
try {} catch (err) {
  const isString = typeof err === "string";
  err = 0;
  if (isString) {
    err;
  }
}
```

Reply:
@andrewbranch Yes, can be. This is a bit tricky case, with a catch clause. Currently, CFA supports only constants.

```ts
declare const a;
const isString = typeof a === "string";
if (isString) {
    a; // string
}
```

```ts
declare let a;
const isString = typeof a === "string";
if (isString) {
    a; // any
}
```

```ts
declare let a;
const isString = typeof a === "string";
a = 1;
if (isString) {
    a; // still any
}
```

I don't know what the best way to resolve this issue. What do you think?

Reply:
I read `isParameter(decl) && !isParameterAssigned(symbol)` to mean that CFA also works for function parameters, which feel very similar to catch clause variables, as long as they’re not reassigned anywhere in the function body. It would be nice if catch clause variables worked the same way as function parameters, but I don’t know how easy it would be to reuse `isParameterAssigned` for catch clauses.

Reply:
Oke, I'll check, thanks